src/storage/temporary_file_manager.cpp

```diff
@@ -4,11 +4,29 @@
 
 namespace duckdb {
 
+//===--------------------------------------------------------------------===//
+// FileSizeMonitor
+//===--------------------------------------------------------------------===//
+
+FileSizeMonitor::FileSizeMonitor(TemporaryFileManager &manager) : manager(manager) {
+}
+
+void FileSizeMonitor::Increase(idx_t blocks) {
+	auto size_on_disk = blocks * TEMPFILE_BLOCK_SIZE;
+	manager.IncreaseSizeOnDisk(size_on_disk);
+}
+
+void FileSizeMonitor::Decrease(idx_t blocks) {
+	auto size_on_disk = blocks * TEMPFILE_BLOCK_SIZE;
+	manager.DecreaseSizeOnDisk(size_on_disk);
+}
+
 //===--------------------------------------------------------------------===//
 // BlockIndexManager
 //===--------------------------------------------------------------------===//
 
-BlockIndexManager::BlockIndexManager() : max_index(0) {
+BlockIndexManager::BlockIndexManager(unique_ptr<FileSizeMonitor> file_size_monitor)
```

Comment:
Can this directly use `TemporaryFileManager &` instead of this file size monitor class? This can then be stored in an `optional_ptr<TemporaryFileManager>`. We can then add two methods:

```cpp
void BlockIndexManager::IncreaseBlockCount(idx_t blocks) {
   if (!manager) {
      return;
   }
   auto size_on_disk = blocks * TEMPFILE_BLOCK_SIZE;
   manager->IncreaseSizeOnDisk(size_on_disk);
}
```